### DISCUSSION TITLE
Backups with k10

### DISCUSSION THREAD

**Discussion by aDingil** [mysticaltech/terraform-hcloud-kube-hetzner#765] [Q&A] [answered]
What do you think about k10.
The installation was surprisingly easy, just skip their marketing:
https://docs.kasten.io/install/other/other.html
Also their pricing is okay i guess.
https://www.kasten.io/pricing

Are there similar concept where to backup pod state and volumes?

**Comment by mysticaltech** [comment, marked as answer]
@aDingil Best to use VMWare [Velero](https://github.com/vmware-tanzu/velero) for general backup, and Longhorn backups for volumes. Many of our users use both and are super happy!

**Reply by aDingil** [reply]
Okay will do.

**Reply by aDingil** [reply]
So i tried longhorn and velero:
longhorn
  - no existing pvc are showing up in the volumes, so i cant backup pvcs
    - Possible reasons
      - misconfiguration by mypart
      - no support for kubernetes >1.26
  - seemingly no support for gcp?!

velero
  - backup/restore seems to work for everything, except ...
  - backing up pvc is "working", but
    - restored pvc seemed to be empty or misconfigured so that pods arent able to access data
      - tested with vaultwarden and crowdsec
    - different install flags for velero doesnt change the above described effect

-> no backup solution seems to work properly?!
I wish i had the couple of thousand buckets extra to use k10 haha (hoping it will solve all my problems).

So want is you specific backup strategy and how did you installed/implemented it?
And what am i doing wrong?

**Reply by mysticaltech** [reply]
@aDingil Most definitely your kube version is too recent, try again with 1.25.

**Reply by mysticaltech** [reply]
Between you can downgrade easily, just edit the system upgrade controller plans. Or even easier, set initial_k3s_version and apply again. Then the downgrade will happen node by node.

**Reply by aDingil** [reply]
Okidoki will report when i fixed my velero pvc problem

**Reply by aDingil** [reply]
so fixed my velero pvc problems, i can confirm after a complete destruction of vaultwarden or crowdsec a complete backup works!
It was a permission issue and rtfm on my part:

# Instructions for gcp:
## Installation
1. `velero install --provider gcp --plugins velero/velero-plugin-for-gcp --bucket {BUCKET_NAME] --secret-file {SECRET} --use-node-agent --default-volumes-to-fs-backup`
2. Then add this to the velero node-agent daemonset
```yaml
spec:
  template:
    spec:
      containers:
      - name: node-agent
        securityContext:
          privileged: true
```

## Backup Creation
`velero backup create {BACKUP_NAME} --include-namespaces {NAME_SPACE}`

## Restore
`velero restore create --from-backup {BACKUP_NAME}`

## NOTE
There still some [limitations](https://velero.io/docs/v1.10/file-system-backup/#limitations), i hope i dont run into them.

**Reply by mysticaltech** [reply]
@aDingil Thanks for sharing, awesome!

**Comment by aDingil** [comment]
a drowback is that with ha you would easily reach the 5 node limit

**Reply by mysticaltech** [reply]
Just shoot a support request to Hetzner, they'll bump it up for you.

**Reply by mysticaltech** [reply]
Also, why is node count related to this? Backups are not directly tied to that, or am I missing someting?

**Reply by aDingil** [reply]
No i meant as a drawback of k10.
Their free tier only allows 5 nodes and as far as my understanding of ha goes, you need >6 nodes.

**Reply by aDingil** [reply]
also why is longhorn needed?
Velero supports also volume backups or am i missing sth?

**Reply by mysticaltech** [reply]
Yes, Velero supports volume backups from my understanding. But if you are using volumes, you are probably also using Longhorn (otherwise I would recommend it, because it's more powerful and stable than Hetzner CSI it seems), and since it has backups integrated too, you can use both or just that, up to you!
